### PR TITLE
VCDA-395 and VCDA-396: Add support for listing and deleting org vdc networks

### PR DESCRIPTION
**Pull request by rocknes**
* Added functionality to get all org vdc network records
* Added functionality to get org vdc network resource, filtered by name and type
* Added functionality to delete org vdc networks by name
* Unit tests for implemented features
* Minor text fixes

Unit test output
---------------------
(pyvcloud) C:\code\pyvcloud\tests>python -m unittest vcd_network.TestNetwork
.......
----------------------------------------------------------------------
Ran 7 tests in 117.558s

OK

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/184)

### CODE REVIEW

**Comment by pacogomez** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pyvcloud/vcd/vdc.py, line 935 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J2CR2M7LmZF6X-dxX:-L3J2CR2M7LmZF6X-dxY:b7ropwm) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L935)):*
> ```Python
>             for record in records.OrgVdcNetworkRecord:
>                 result.append(record)
>         return result
> ```

would it be ok to just to return `records.OrgVdcNetworkRecord`, the result of the `get_linked_resource` call?

---

*[pyvcloud/vcd/vdc.py, line 966 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J22C5tkFtvD490BuW:-L3J22C6kVzpEFZ7RmGS:b-6wovzk) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L966)):*
> ```Python
>                     continue
>             result.append(orgvdc_network_resource)
>         return result
> ```

why is it not case sensitive?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by rocknes** [comment]
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pyvcloud/vcd/vdc.py, line 935 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J2CR2M7LmZF6X-dxX:-L3QBku71nrpv0S8T5P0:bjl1hi5) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L935)):*
<details><summary><i>Previously, pacogomez (Paco Gómez) wrote…</i></summary><blockquote>

would it be ok to just to return `records.OrgVdcNetworkRecord`, the result of the `get_linked_resource` call?
</blockquote></details>

No, the xml returned by vcd also has \<Link\> elements, we need to filter those out

---

*[pyvcloud/vcd/vdc.py, line 966 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J22C5tkFtvD490BuW:-L3QBx-ug74W6hL9Q96e:b-u81ozl) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L966)):*
<details><summary><i>Previously, pacogomez (Paco Gómez) wrote…</i></summary><blockquote>

why is it not case sensitive?
</blockquote></details>

vCD in general is case insensitive.
I am sure that org names doesn't care about case.
I can verify if that's the case with org networks too.  If it turns out otherwise I will make the check case sensitive.

I believe we have a task in the backlog to make our cli case insensitive, this is more like preemptive work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by sahithi** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[pyvcloud/vcd/vdc.py, line 935 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J2CR2M7LmZF6X-dxX:-L3UriJA889b14sYmD2z:b-asbks1) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L935)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

No, the xml returned by vcd also has \<Link\> elements, we need to filter those out
</blockquote></details>

+1 Paco 

if hasattr(records, 'OrgVdcNetworkRecord'):
          return records.OrgVdcNetworkRecord

---

*[pyvcloud/vcd/vdc.py, line 974 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UvGpvfzscbV35Ivkg:-L3UvGpwLqGWZJFS2Fdp:bcfq37) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L974)):*
> ```Python
>             representing all directly connected orgvdc network resources.
>         """
>         return self._get_orgvdc_networks(type='bridged')
> ```

Are we going to have more of these constant values like 'bridged', 'isolated'? I see 4 places you've used these currently. If yes, consider a re-design with global constant variables or enum in one of your upcoming tasks.

---

*[pyvcloud/vcd/vdc.py, line 1029 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UxF5ISkzpMpRW-9gG:-L3UxF5ISkzpMpRW-9gH:bm3rk0) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1029)):*
> ```Python
>         :raises: Exception: If orgvdc network with the given name is not found.
>         """
>         net_resource = self.get_direct_orgvdc_network(name)
> ```

To delete a network all we need is 'href'. self.get_direct_orgvdc_network(name) loops through records getting resource objects for each record. Seems redundant. However, I see that you are specifically looking for one of direct or isolated networks. That raises another question on whether ovdc allows to have direct and isolated network with the same name? 
If yes, 
      skip my comment, 
else
     We can get href from the record itself. There would be no need for two separate methods.
     (Please add a TODO for now if you are busy finishing other pending tasks for RC)

---

*[pyvcloud/vcd/vdc.py, line 1046 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UzSMLIuj3Y32-jDmt:-L3UzSMLIuj3Y32-jDmu:bj81f4e) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1046)):*
> ```Python
>         :raises: Exception: If orgvdc network with the given name is not found.
>         """
>         net_resource = self.get_isolated_orgvdc_network(name)
> ```

Refer delete method above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by rocknes** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[pyvcloud/vcd/vdc.py, line 935 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J2CR2M7LmZF6X-dxX:-L3VYjbatLtMffDEywlz:b-5rk9tw) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L935)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

+1 Paco 

if hasattr(records, 'OrgVdcNetworkRecord'):
          return records.OrgVdcNetworkRecord
</blockquote></details>

Got it. You guys are right, my bad fixing it.

---

*[pyvcloud/vcd/vdc.py, line 974 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UvGpvfzscbV35Ivkg:-L3VYnNFIQ-Sh3dFOn_m:bp33nfa) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L974)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Are we going to have more of these constant values like 'bridged', 'isolated'? I see 4 places you've used these currently. If yes, consider a re-design with global constant variables or enum in one of your upcoming tasks.
</blockquote></details>

Sounds like a good idea, will introduce the enum in a separate commit.

---

*[pyvcloud/vcd/vdc.py, line 1029 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UxF5ISkzpMpRW-9gG:-L3VYwdWLcGqfMpQiDdr:b-r8ht44) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1029)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

To delete a network all we need is 'href'. self.get_direct_orgvdc_network(name) loops through records getting resource objects for each record. Seems redundant. However, I see that you are specifically looking for one of direct or isolated networks. That raises another question on whether ovdc allows to have direct and isolated network with the same name? 
If yes, 
      skip my comment, 
else
     We can get href from the record itself. There would be no need for two separate methods.
     (Please add a TODO for now if you are busy finishing other pending tasks for RC)
</blockquote></details>

No, two nets, a direct and an isolated network can't have the same name in a given ovdc. But if we just use the records to figure out the href, we run the risk of blurring the boundaries between isolated and direct networks. Let me give an example,

If we use records to figure out the href, then we are not filtering on the type of network, since record doesn't have that info. As a result the following two commands will have the same effect

vcd network direct delete direct-net1
vcd network isolated delete direct-net1

In essence we are letting the user delete a direct network using commands meant for isolated networks.

---

*[pyvcloud/vcd/vdc.py, line 1046 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UzSMLIuj3Y32-jDmt:-L3VZuEoXha3tGdBpDRb:b-wfxxxd) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1046)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Refer delete method above
</blockquote></details>

Answered above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by rocknes** [comment]
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[pyvcloud/vcd/vdc.py, line 966 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3J22C5tkFtvD490BuW:-L3VlN98pxQBKh59Ld8Q:bm3oael) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L966)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

vCD in general is case insensitive.
I am sure that org names doesn't care about case.
I can verify if that's the case with org networks too.  If it turns out otherwise I will make the check case sensitive.

I believe we have a task in the backlog to make our cli case insensitive, this is more like preemptive work.
</blockquote></details>

I checked it and it seems that name of networks are case sensitive. Removing the case insensitive checks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by sahithi** [comment]
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pyvcloud/vcd/vdc.py, line 974 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UvGpvfzscbV35Ivkg:-L3ZQF0djzrS6sJLLrGX:b-v4gpik) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L974)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Sounds like a good idea, will introduce the enum in a separate commit.
</blockquote></details>

Please add a TODO here for now.

---

*[pyvcloud/vcd/vdc.py, line 1029 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UxF5ISkzpMpRW-9gG:-L3ZIsTigJPtxwWeBxei:bmi1g79) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1029)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

No, two nets, a direct and an isolated network can't have the same name in a given ovdc. But if we just use the records to figure out the href, we run the risk of blurring the boundaries between isolated and direct networks. Let me give an example,

If we use records to figure out the href, then we are not filtering on the type of network, since record doesn't have that info. As a result the following two commands will have the same effect

vcd network direct delete direct-net1
vcd network isolated delete direct-net1

In essence we are letting the user delete a direct network using commands meant for isolated networks.
</blockquote></details>

Okay, in my opinion, "vcd network delete <network name>" at one level up should do the job efficiently.

At the same time, I don't mind user having extra clarity on what type of network they are just about to delete, so you can keep the commands (vcd network direct delete & isolated delete)  as-is.

To sum up, I think we should have another delete command one level up
vcd network delete <network-name>

[I have added  a comment on networking confluence page to add two more commands at 'network' level
vcd network list // Lists tuple (name, type) of all networks in the org
vcd network delete network-name]

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by sahithi** [comment]
:lgtm: (Add TODO for enums)

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-:-L3V-GhCpL_0obeMsQbW:bhz5s9c)*

**Comment by rocknes** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pyvcloud/vcd/vdc.py, line 1029 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UxF5ISkzpMpRW-9gG:-L3Zbo8XufN5YI5L7qNw:b-uzreh8) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1029)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Okay, in my opinion, "vcd network delete <network name>" at one level up should do the job efficiently.

At the same time, I don't mind user having extra clarity on what type of network they are just about to delete, so you can keep the commands (vcd network direct delete & isolated delete)  as-is.

To sum up, I think we should have another delete command one level up
vcd network delete <network-name>

[I have added  a comment on networking confluence page to add two more commands at 'network' level
vcd network list // Lists tuple (name, type) of all networks in the org
vcd network delete network-name]


</blockquote></details>

External network can be namesake to an orgvdc network, so having a global delete command viz.
vcd network delete <network-name> without specifying the type is not feasible. And if we have to specify the type of network, then we are better off using the sub command pattern.

External networks are in a different league than org vdc networks. I don't see a reason to list them together. Also, the implementation will be murky if we try to bunch them up, since 
1. external networks are only accessible by sys admins, i.e. our implementation will need to have user centric logic or else any non sys admin will get exception while listing networks
2. there is no direct query to list external networks and orgvdc networks in one shot, i.e. we will have to call two different functions and combine their outputs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by hodgesrm** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pyvcloud/vcd/vdc.py, line 735 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3ZimcLrS5r7D7v9Kq6:-L3ZimcMd_wPHxhdsbwX:b-3k8l8l) ([raw file](https://github.com/vmware/pyvcloud/blob/fa2bc5ae4e2eb7e1235ea08458dd7012dcb86d19/pyvcloud/vcd/vdc.py#L735)):*
> ```Python
>         :param name: (str) Name of the new vApp.
>         :param description: (str) Description of the new vApp.
>         :param network: (str) Name of the OrgVDC network, the vApp will
> ```

'network," should be just "network" (no comma)

---

*[pyvcloud/vcd/vdc.py, line 936 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3ZkG9DO1TcIjDZpIE7:-L3ZkG9DO1TcIjDZpIE8:b-cl5xpn) ([raw file](https://github.com/vmware/pyvcloud/blob/fa2bc5ae4e2eb7e1235ea08458dd7012dcb86d19/pyvcloud/vcd/vdc.py#L936)):*
> ```Python
>             return []
> 
>     def _get_orgvdc_networks(self, name=None, type=None):
> ```

Why not make this public?  It's useful. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by rocknes** [comment]
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pyvcloud/vcd/vdc.py, line 974 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UvGpvfzscbV35Ivkg:-L3Zl9rZXOigOSFz6UYw:bu7d8b5) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L974)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Please add a TODO here for now.
</blockquote></details>

Added the enums :)

---

*[pyvcloud/vcd/vdc.py, line 936 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3ZkG9DO1TcIjDZpIE7:-L3ZmAXEOEvKVQ7uegzA:b-80tlqk) ([raw file](https://github.com/vmware/pyvcloud/blob/fa2bc5ae4e2eb7e1235ea08458dd7012dcb86d19/pyvcloud/vcd/vdc.py#L936)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

Why not make this public?  It's useful. 
</blockquote></details>

Sure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by pacogomez** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by hodgesrm** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-:-L3Zrav0LqRcduI3VQhC:bnfp4nl)*

**Comment by pacogomez** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pyvcloud/vcd/vdc.py, line 1029 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-L3UxF5ISkzpMpRW-9gG:-L3ZqbNMQMXySW_PiBIG:b-etr2qk) ([raw file](https://github.com/vmware/pyvcloud/blob/9d665f52e97ba5e87423ff44eba3e471262c2457/pyvcloud/vcd/vdc.py#L1029)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

External network can be namesake to an orgvdc network, so having a global delete command viz.
vcd network delete <network-name> without specifying the type is not feasible. And if we have to specify the type of network, then we are better off using the sub command pattern.

External networks are in a different league than org vdc networks. I don't see a reason to list them together. Also, the implementation will be murky if we try to bunch them up, since 
1. external networks are only accessible by sys admins, i.e. our implementation will need to have user centric logic or else any non sys admin will get exception while listing networks
2. there is no direct query to list external networks and orgvdc networks in one shot, i.e. we will have to call two different functions and combine their outputs.



</blockquote></details>

I think we should keep the sub-command scheme to work with different types of networks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184#-:-L3ZrleoVipgnPXlSsqq:bnfp4nl)*

**Comment by pacogomez** [comment]
Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*

**Comment by pacogomez** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/184)*
